Api/edit rif endpoint

BonfaceKilz opened the pull request:
## Description

This PR extends the current RIF edit end-points to allow editing in RDF.

BonfaceKilz committed:
Run python-black against file.

Signed-off-by: Munyoki Kilyungi <me@bonfacemunyoki.com>

BonfaceKilz committed:
Add new method for updating RDF Graph content.

* gn3/db/rdf/__init__.py: Import POST, DIGEST, JSON.
(update_rdf): New function.

Signed-off-by: Munyoki Kilyungi <me@bonfacemunyoki.com>

BonfaceKilz committed:
Rename sparql_construct_query -> sparql_query.

* gn3/db/rdf/__init__.py (sparql_construct_query): Rename to...
(sparql_query): ... this.  Extend function to be able to return JSQN
for SELECT queries.
(query_frame_and_compact): Ditto.
(query_and_compact): Ditto.
(query_and_frame): Ditto.

Signed-off-by: Munyoki Kilyungi <me@bonfacemunyoki.com>

BonfaceKilz committed:
Fetch the next comment version in RDF.

* gn3/db/rdf/wiki.py (get_next_comment_version): New function.

Signed-off-by: Munyoki Kilyungi <me@bonfacemunyoki.com>

BonfaceKilz committed:
Fetch symbol by it's name.

The RDF ttl files were updated to have:

comment gnt:symbol "name" .

instead of:

comment gnt:symbol symbolQname .

* gn3/db/rdf/wiki.py (get_wiki_entries_by_symbol): Fetch symbol using
it's string literal name.
(get_comment_history): Ditto.

Signed-off-by: Munyoki Kilyungi <me@bonfacemunyoki.com>

BonfaceKilz committed:
Add method that inserts new RIF comments.

* gn3/db/rdf/wiki.py: Import datetime, update_rdf.
(update_wiki_comment): New function.

Signed-off-by: Munyoki Kilyungi <me@bonfacemunyoki.com>

BonfaceKilz committed:
Fix typo.

Signed-off-by: Munyoki Kilyungi <me@bonfacemunyoki.com>

BonfaceKilz committed:
Update "POST /wiki/comment_id/edit" to have RDF edits.

* gn3/api/metadata_api/wiki.py: Import update_wiki_comment.
(edit_wiki): Update wiki in RDF.

Signed-off-by: Munyoki Kilyungi <me@bonfacemunyoki.com>

BonfaceKilz committed:
Update docs with extra conf vars for updating virtuoso.

Signed-off-by: Munyoki Kilyungi <me@bonfacemunyoki.com>

jnduli commented on README.md:
```diff
@@ -124,6 +124,10 @@ These configurations should be set in an external config file, pointed to with t
 - XAPIAN_DB_PATH (ex: "/export/data/genenetwork/xapian")
 - TMPDIR
+- SPARQL_USER
+- SPARQL_ENDPOINT (ex: "http://localhost:9082/sparql-auth/")
```
- `SPARQL_AUTH_ENDPOINT` because we already have another `SPARQL_ENDPOINT`?
- does sparql endpoint support having credentials in the url e.g. `http://username@password:localhost....` ref https://developer.mozilla.org/en-US/docs/Web/HTTP/Authentication#access_using_credentials_in_the_url (If this works it reduces the number of config variables)

BonfaceKilz replied:
That `SPARQL_AUTH_ENDPOINT` because that is what Virtuoso provides for INSERTS/UPDATES/DELETES operations in a graph.  To use that end-point, you have to provide a user name and a password.

jnduli commented on gn3/db/rdf/wiki.py:
```diff
@@ -180,3 +187,112 @@ def get_comment_history(
     return results
+
+
+def get_next_comment_version(
```
what should be the source of truth for this function? There's a similar function for this [here](https://github.com/jnduli/genenetwork3/blob/52cd294c2f1d06dddbd6ff613b11f8bc43066038/gn3/db/wiki.py#L51) and I'd rather we stick to one. This prevents drifts when problems occur.

jnduli commented on gn3/db/rdf/wiki.py:
```diff
@@ -180,3 +187,112 @@ def get_comment_history(
+        comment=payload["comment"],
+        name=name, symbol=payload['symbol'],
+        comment_id=comment_id, next_version=next_version,
+        created=datetime.datetime.now(datetime.timezone.utc).strftime('%Y-%m-%d %H:%M:%S'))
```
this will be different from the value we store in mysql. Can we make it the same value?

jnduli commented on gn3/api/metadata_api/wiki.py:
```diff
@@ -66,6 +70,18 @@ def edit_wiki(comment_id: int):
+        # Editing RDF:
+        update_wiki_comment(
+            comment_id=comment_id,
+            payload=payload,
```
Can we pass over `insert_dict` instead to this child process? I don't like passing in the payload since `insert_dict` would be what we want to insert after doing clean up. We can modify it to work both in mysql and in rdf

BonfaceKilz replied:
I've had to add extra keys to the `insert_dict`.

jnduli commented on gn3/db/rdf/wiki.py:
```diff
@@ -180,3 +187,112 @@ def get_comment_history(
+        sparql_password=sparql_conf["sparql_password"],
+        sparql_auth_uri=sparql_conf["sparql_auth_uri"],
+    )
+    return (res, 200)
```
why are we returning 200 here?

BonfaceKilz replied:
No good reason.  I reckon we can just return the `res` only.

jnduli commented on gn3/db/rdf/__init__.py:
```diff
@@ -126,66 +127,88 @@
+)
+
+
+def sparql_query(query: str, endpoint: str, format_type="json-ld") -> dict:
```
minor question: can we set the format_type to be types supported by sparqlwrapper? I was thinking a uniont type between `JSON | JSONLD`, see https://sparqlwrapper.readthedocs.io/en/latest/SPARQLWrapper.Wrapper.html

BonfaceKilz committed:
Get the next versionId from SQL.

* gn3/api/metadata_api/wiki.py (edit_wiki): Get the next version Id
from SQL
* gn3/db/rdf/wiki.py (get_next_comment_version): Delete.
(update_wiki_comment): Provide the next_version id as an extra arg.

Signed-off-by: Munyoki Kilyungi <me@bonfacemunyoki.com>

BonfaceKilz committed:
Pass in the createtime as an arg when inserting into RDF graph.

* gn3/api/metadata_api/wiki.py (edit_wiki): Pass in the createtime as
an arg.
* gn3/db/rdf/wiki.py: Remove un-used imports.
(update_wiki_comment): Use passed in "created" arg in the createtime
field.  This makes sure that the SQL/RDF timestamps match.

Signed-off-by: Munyoki Kilyungi <me@bonfacemunyoki.com>

BonfaceKilz committed:
Pass in insert_dict instead of payload in "get_wiki_entries".

* gn3/api/metadata_api/wiki.py (edit_wiki) [insert_dict]: Add species,
categories fields.
[update_wiki_comment]: Pass in insert_dict, and get rid of redundant
keys.  Directly pass in sparql_auth's conf params instead of a dict.
* gn3/db/rdf/wiki.py (get_next_comment_version): Get rid of redundant
params.  Use fields directly from insert_dict.  Directly pass in
sparql_auth's conf params instead of a dict.

Signed-off-by: Munyoki Kilyungi <me@bonfacemunyoki.com>
